ocaml/typing/solver.ml

```diff
@@ -0,0 +1,724 @@
+        in terms of submoding, but we have now ensured that a is within the
+        range of f v, and thus a is within the co-domain of f, and thus its
+        safe to apply f' to a. *)
+    if C.le obj a (mlower obj mv)
+    then Ok ()
+    else if not (C.le obj a (mupper obj mv))
+    then Error (mupper obj mv)
+    else
+      let a = C.join obj a (mlower obj mv) in
+      let f' = C.left_adjoint obj f in
+      let src = C.src obj f in
+      let a' = C.apply src f' a in
+      assert (Result.is_ok (submode_cv ?log src a' v));
+      Ok ()
+
+  (* In the following, functions might take two logs: [log] records changes
+     caused by the operation, [optim] records optimizing changes not caused by the
+     operation. Semantically, [optim] happens before [log] and doesn't rely on [log].
+     On the other hand, some changes in [log] might rely on [optim]. *)
+
+  (** Returns [None] if success; [Some x] if failed, and [x] is the next best
+     guess to replace the constant argument that MIGHT succeed. *)
+  let rec submode_vc :
+      type a. ?log:_ -> a C.obj -> a var -> a -> (unit, a) Result.t =
+    fun (type a) ?log (obj : a C.obj) v a' ->
+     if C.le obj v.upper a'
+     then Ok ()
+     else if not (C.le obj v.lower a')
+     then Error v.lower
+     else (
+       update_upper ?log obj v a';
+       let r =
+         v.vlower
+         |> find_error (fun mu ->
+                let r = submode_mvc ?log obj mu a' in
+                (if Result.is_ok r
+                then
+                  (* Optimization: update [v.lower] based on [mlower u].*)
+                  let mu_lower = mlower obj mu in
+                  if not (C.le obj mu_lower v.lower)
+                  then update_lower ?log obj v mu_lower);
+                r)
+       in
+       if C.le obj v.upper v.lower then set_vlower ?log v [];
+       r)
+
+  and submode_mvc :
+        'a 'r.
+        ?log:change list ref ->
+        'a C.obj ->
+        ('a, allowed * 'r) morphvar ->
+        'a ->
+        (unit, 'a) Result.t =
+   fun ?log obj (Amorphvar (v, f) as mv) a ->
+    (* See [submode_cmv] for why we need the following seemingly redundant
+       lines. *)
+    if C.le obj (mupper obj mv) a
+    then Ok ()
+    else if not (C.le obj (mlower obj mv) a)
+    then Error (mlower obj mv)
+    else
+      let a = C.meet obj a (mupper obj mv) in
+      let f' = C.right_adjoint obj f in
+      let src = C.src obj f in
+      let a' = C.apply src f' a in
+      (* cannot assert [Option.is_none] because [mlower] is imprecise. *)
+      Result.map_error (C.apply obj f) (submode_vc ?log src v a')
+
+  (** Calculate the precise lower bound *)
+  let zap_to_floor_try (type a) (obj : a C.obj) (v : a var) =
+    let rec loop lower =
+      let log = ref [] in
+      let r = submode_vc ~log obj v lower in
+      match r with
+      | Ok () -> log, lower
+      | Error a ->
+        undo_changes !log;
+        loop (C.join obj a lower)
```

Comment:
I think one thing that's not obvious (although commented) is that `submode_vc` in case of `Error` gives `a` that's strictly higher than `lower`. This, combined with the fact that our lattices are all finite sets, guarantees termination.

I don't think this generalises to infinite lattices. I will add comments to make it clear.

Reply:
I don't see this fact commented on `submode_vc`. (I kinda see how it's implied by the comments that are there, but it's better to be direct, I think.)

Reply:
This is mentioned in another comment and fixed - resolved.